nevergrad/optimization/optimizerlib.py

```diff
@@ -1203,68 +1203,74 @@ def _internal_tell_candidate(self, candidate: base.Candidate, value: float) -> N
+            self._optimizers_ = []
+            conv = {"num_workers": self.num_workers, "dimension": self.dimension}
+            budgets = [conv[b] if isinstance(b, str) else b for b in self._parameters.budgets]
+            last_budget = None if self.budget is None else self.budget - sum(budgets)
+            for opt, budget in zip(self._parameters.optimizers, budgets + [last_budget]):  # type: ignore
+                self._optimizers_.append(opt(self.instrumentation, budget=budget, num_workers=self.num_workers))
+        return self._optimizers_  # type: ignore
+
+    def _internal_ask_candidate(self) -> base.Candidate:
+        # Which algorithm are we playing with ?
+        sum_budget = 0.0
+        opt = self._optimizers[0]
+        for opt in self._optimizers:
+            sum_budget += float("inf") if opt.budget is None else opt.budget
+            if self.num_ask < sum_budget:
+                break
+        # if we are over budget, then use the last one...
+        return opt.ask()
+
+    def _internal_tell_candidate(self, candidate: base.Candidate, value: float) -> None:
+        # Let us inform all concerned algorithms
+        sum_budget = 0.0
+        for opt in self._optimizers:
+            sum_budget += float("inf") if opt.budget is None else opt.budget
+            if self.num_tell < sum_budget:
+                opt.tell(candidate, value)
+
+
+class Chaining(base.ParametrizedFamily):
     """
     A chaining consists in running algorithm 1 during T1, then algorithm 2 during T2, then algorithm 3 during T3, etc.
-    Each algorithm (except the first one) is fed with the entire archive.
-    """
-    
-    def __init__(
-        self,
-        list_of_algorithms,
-        list_of_budgets,
-    ) -> None:
-        # Either we have the budget for each algorithm, or the last algorithm uses the rest of the budget, so:
-        assert len(list_of_algorithms) == len(list_of_budgets) or len(list_of_algorithms) == len(list_of_budgets) + 1 
+    Each algorithm is fed with what happened before it.
 
-        # Let us store the two key parameters:
-        self._list_of_budgets = list_of_budgets
-        self._list_of_algorithms = list_of_algorithms
+    Parameters
+    ----------
+    optimizers: list of Optimizer classes
+        the sequence of optimizers to use
+    budgets: list of int
+        the corresponding budgets for each optimizer but the last one
 
+    """
+    _optimizer_class = _Chain
+
+    def __init__(self, optimizers: Sequence[Union[base.OptimizerFamily, Type[base.Optimizer]]],
+                 budgets: Sequence[Union[str, int]]) -> None:
+        # Either we have the budget for each algorithm, or the last algorithm uses the rest of the budget, so:
+        self.budgets = tuple(budgets)
+        self.optimizers = tuple(optimizers)
+        assert len(self.optimizers) == len(self.budgets) + 1
+        assert all(x in ("dimension", "num_workers") or x > 0 for x in self.budgets)  # type: ignore
```

Comment:
i've changed to string, it's easier to read

Reply:
Ok cleaner than my jurassic-style version :-)